File: CodingSamples/DotNet/Fundamentals/Inheritance/InterfaceTypeTest/Banking/Account.cs
namespace Banking;

//a class defined with public keyword is visible
//to other referencing assemblies/projects
//a class defined with abstract keyword does not
//support activation, it can only be used as a 
//base class for other classes
public abstract class Account
{
    public long Id { get; internal set; }

    public decimal Balance { get; protected set; }

    //a method declared with abstract keyword is a pure virtual
    //instance method, it can only appear in an abstract class and
    //it must be overridden by a non-abstract derived class 
    public abstract void Deposit(decimal amount);

    public abstract void Withdraw(decimal amount);

    public void Freeze()
    {
        Id = -Id;
    }
}

File: CodingSamples/Java/Fundamentals/Inheritance/InterfaceTypeTest/bank/banking/Account.java
package banking;

//a class defined with abstract keyword cannot be instantiated
public abstract class Account {
    
    long id;
    protected double balance;

    public long id() {
        return id;
    }

    public double balance() {
        return balance;
    }

    //a method declared with abstract keyword
    //must be overridden in the subclass
    public abstract void deposit(double amount);

    public abstract void withdraw(double amount) throws InsufficientFundsException;

    //a method declared with final keyword cannot be overridden
    public final void freeze() {
        id = -id;
    }
}

